### PR TITLE
[SYCL][E2E] Add a test to track UNSUPPORTED tests

### PR DESCRIPTION
**Pull request by KornevNikita**
Similar to XFAIL tests (no-xfail-without-tracker.cpp) adding a requirement to follow UNSUPPORTED by UNSUPPORTED-TRACKER.

Adding a test to track the amount and the list of improperly UNSUPPORTED tests.

Also update some tests to check the script.

### CODE REVIEW

**Comment by KornevNikita** [comment]
@steffenlarsen @AlexeySachkov I know you suggested "UNSUPPORTED-REASON", but if you don't mind I'd prefer "UNSUPPORTED-TRACKER" to unify this with XFAIL, so we don't have to remember what and where we should use. Or any other unified mark.

File: sycl/test-e2e/Basic/kernel_max_wg_size.cpp
Comment (steffenlarsen) [inline sycl/test-e2e/Basic/kernel_max_wg_size.cpp:8]: If we're going to call it `UNSUPPORTED-TRACKER`, this sort of case seems weird. Could we then name it `UNSUPPORTED-INTENDED` for these cases and a case of `UNSUPPORTED` must have one or the other following it? I.e.
```suggestion
// UNSUPPORTED-INTENDED: This extension is currently implemented in DPC++ only for Intel GPU devices and only when using the Level Zero backend or OpenCL backend.
```

Comment (KornevNikita) [reply sycl/test-e2e/Basic/kernel_max_wg_size.cpp]: That'd better. 7974dbf

**Comment by KornevNikita** [comment]
NOTE: need to re-run pre-commit before merging.

**Review by steffenlarsen — APPROVED** [review]
Let's get it running!

File: sycl/test-e2e/Basic/kernel_max_wg_size.cpp
Comment (AlexeySachkov) [inline sycl/test-e2e/Basic/kernel_max_wg_size.cpp:8]: Unrelated to the main change in the PR, but tagging @sarnex here. I'm not sure if having an optional property unguarded by an aspect is a good interface for end users. Even if we don't introduce an aspect, can we update the spec to say that submitting a kernel with `grf` property to a device and/or using a backend which doesn't support that results in an exception? This way we would be able to rewrite the test to avoid use of `UNSUPPORTED.

Moreover, we have `REQUIRES` which points to specific GPUs, do we even need `UNSUPPORTED` in that case?

File: sycl/test-e2e/README.md
Comment (dm-vodopyanov) [inline sycl/test-e2e/README.md:366]: Is it an issue number from the internal tracking system?

File: sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp
Comment (AlexeySachkov) [inline sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp:35]:
```suggestion
// RUN: grep -Pv "UNSUPPORTED-TRACKER:\s+(?:https://github.com/[\w\d-]+/[\w\d-]+/issues/[\d]+)|(?:[\w]+-[\d]+)|(?:UNSUPPORTED-INTENDED:\s\w.*)" > %t
```

- to require nice formatting (i.e. exactly one spacing symbol after `-INDENDED:`
- to require at least single alphanumeric symbol as an explanation. Not that it would prevent all possible malformed explanations, but that's at least something.

But that is mostly a minor comment, I believe that formatting and reasoning will be up to reviewers anyways.

File: sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp
Comment (AlexeySachkov) [inline sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp:33]: `.c` include is a certain legacy from `XFAIL` test: I'm not entirely sure if we ever should have any `.c` tests. SYCL is C++17

That's minor, we can address it in a follow-up PR for both `XFAIL` and `UNSUPPORTED`

File: sycl/test-e2e/README.md
Comment (KornevNikita) [reply sycl/test-e2e/README.md:366]: Yes, now we allowed to expose it. But no links :)

**Review by AlexeySachkov — APPROVED** [review]
I still prefer `-REASON`, but that's not a strong enough objection to block the PR.

I think it is good to be merged. @KornevNikita, could you please do a local merge + test and if everything is fine, let me know and I will press the merge button (we don't really need to re-run the whole CI for this PR).

File: sycl/test-e2e/README.md
Comment (dm-vodopyanov) [reply sycl/test-e2e/README.md:366]: I'm not sure we can still do that. Can you please consult with our security team, or there was such conversation already?

File: sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp
Comment (KornevNikita) [reply sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp:33]: Looks like there are no `.c` tests in test-e2e at all.

**Comment by AlexeySachkov** [comment]
The update to the PR is very small (one more test was recorded as improperly unsupported), @KornevNikita verified it locally and I don't think that it is reasonable to wait for the whole pre-commit and risk the new test being broken again by some other PRs going in. Therefore, I plan to cancel the remaining of the pre-commit and proceed with the merge.

@KornevNikita, please monitor `sycl` branch state after the merge (i.e. post-commit) to quickly send any follow-up PRs if we happen to have conflicts with other PRs merged whilst I was writing this comment

**Comment by AlexeySachkov** [comment]
Also tagging @intel/llvm-gatekeepers here for awareness: any PR which deals with `UNSUPPORTED` directives and had passed pre-commit before this change was merged has to be re-tested, or otherwise we will see failures in both post and pre-commits.

Those failures will be easy to address, but still pinging you explicitly for awareness.

File: sycl/test-e2e/README.md
Comment (AlexeySachkov) [reply sycl/test-e2e/README.md:366]: > I'm not sure we can still do that. Can you please consult with our security team, or there was such conversation already?

I've sent to you an internal wiki. We have indeed got an approval to mention IDs of internal trackers in public.

File: sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp
Comment (AlexeySachkov) [reply sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp:33]: > Looks like there are no `.c` tests in test-e2e at all.

Yeah, there shouldn't be. I probably followed `suffixes` which we have listed in `lit.cfg.py`. In any case I don't think that it is possible to use `-fsycl` in C mode, or use SYCL headers in `.c` mode, so I think that we can do a separate cleanup everywhere to explicitly say that `.c` is not considered as a test for us and drop it from XFAIL/UNSUPPORTED checks

File: sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp
Comment (sarnex) [inline sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp:26]: i didnt actually review the code but i can say it looks correct to run in check-sycl, thanks!

**Comment by sarnex** [comment]
IMO we should fix all of those to use `UNSUPPORTED `and if it's possible add a test to verify anything unsupported is using `UNSUPPORTED`

**Comment by KornevNikita** [comment]
> Hey, just wanted to ask if there are plans to deal with other ways that tests are currently marked as unsupported not directly through the unsupported keyword. The two cases I am aware of would be, requiring a feature that doesn't exist (in the cases i've seen `TEMPORARY_DISABLED`), or misspelling `RUN:` (as `RUNx:` in the examples i've seen) so that certain lines are not picked up.
> 
> Some examples:
> 
> * [`Basic/fpga_tests/fpga_io_pipes.cpp`](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Basic/fpga_tests/fpga_io_pipes.cpp#L3) This test compiles but never runs because of `RUNx:`
> * [`SubGroup/generic_reduce.cpp`](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/SubGroup/generic_reduce.cpp#L8) This test uses `REQUIRES: TEMPORARY_DISABLED`

RIght, we plan to update TEMPORARY_DISABLED tests to use UNSUPPORTED. But currently `UNSUPPORTED: *` results in failed test, so need to update this behavior first.

**Comment by bader** [comment]
> * [`Basic/fpga_tests/fpga_io_pipes.cpp`](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Basic/fpga_tests/fpga_io_pipes.cpp#L3) This test compiles but never runs because of `RUNx:`

I suggest removing the `RUNx:` command in this specific test.

**Pull request merged**
